### PR TITLE
Documentation: preprocessing

### PR DESCRIPTION
**Pull request by dixneuf19**

### CODE REVIEW

File: Notebooks/construct_index.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import argparse
```
Comment (Antoine-marchais) [inline Notebooks/construct_index.py]: Why in Notebooks ?

File: src/interface.py [hunk 2]
```diff
@@ -37,14 +38,14 @@ def retrieve_docs_from_vectorial_query(query: str, inverted_index: InvertedIndex
     parser = argparse.ArgumentParser()
     parser.add_argument("--model", help="<boolean|vectorial> model used to process query", default="boolean")
     parser.add_argument("query", help="query to process")
-    parser.add_argument("--number", "-n", help="number of results (only for boolean)", type=int, default=10)
```
Comment (Antoine-marchais) [inline src/interface.py:40]: I don't think there is a notion of ranking for boolean queries. If there isn't, it makes no sense to truncate the results

File: src/interface.py [hunk 3]
```diff
@@ -23,10 +24,10 @@ def retrieve_docs_from_bool_query(query: str, inverted_index: InvertedIndex) ->
     postfix_query = bq.query_to_postfix(lemmatized_query)
     relevant_documents_id = bq.process_postfix_query(postfix_query, inverted_index.index)
 
-    return [inverted_index.mapping[doc_id] for doc_id in relevant_documents_id]
+    return [inverted_index.mapping[doc_id] for doc_id in relevant_documents_id][:n_results]
```
Comment (Antoine-marchais) [inline src/interface.py]: no ordering with boolean queries

File: src/utils.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import functools
+import logging
+import time
+
+def timer(func):
```
Comment (Antoine-marchais) [inline src/utils.py:5]: Although it is prettier, I feel like this was already done by tqdm in most cases

File: src/interface.py (hunk 2 above)
Comment (Antoine-marchais) [reply src/interface.py:40]: that's actually a typo of mine, I meant to write "only for vectorial"

File: src/preprocess.py [hunk 5]
```diff
@@ -152,10 +157,19 @@ def remove_stop_words_collection(collection: OrdDict[str, List[str]], stop_word_
         {OrdDict[str, List[str]]} -- updated corpus
     """
     stp = load_stop_words(stop_word_path)
+    init_coll_size = get_collection_size(collection)
     new_corpus = OrderedDict()
     for key in tqdm(collection, desc="removing stop words"):
         new_corpus[key] = remove_stop_words_from_document(collection[key], stp, [])
-    return new_corpus 
+    end_coll_size = get_collection_size(new_corpus)
```
Comment (Antoine-marchais) [inline src/preprocess.py:164]: This part of the analysis was already done in my notebook. If we really want metrics on corpus reduction at each step, you might as well write another decorator that monitors number of token and vocab size and use it on stop words, filtering and lemmatization.

File: src/preprocess.py [hunk 6]
```diff
@@ -262,14 +285,15 @@ class InvertedIndex:
     """
     itype: int
     index: Union[
-        OrdDict[str, List[int]], # itype == 1
+        OrdDict[str, OrdDict[int, bool]], # itype == 1
```
Comment (Antoine-marchais) [inline src/preprocess.py:288]: I guess the ultimate optimization for the inner data structure would be a numpy array given the number of documents. Given the size of the vocabulary of the collection, you can even have a 2D array where rows are ordered by alphabetic order on the vocab. But hey if it works, that's fine with me.

File: src/preprocess.py [hunk 7]
```diff
@@ -262,14 +285,15 @@ class InvertedIndex:
...
         OrdDict[str, OrdDict[int, int]], # itype == 2
         OrdDict[str, OrdDict[int, List[int]]] # itype == 3
         ]
     mapping: OrdDict[int, str]
     stats: StatCollection
 
 # TODO: @Antoine-marchais : This part could be refactored with auxiliary function for lisibility.
+@timer
 def build_inverted_index(
```
Comment (Antoine-marchais) [inline src/preprocess.py:297]: Still no factorization there, you deserved it : 
image

File: README.md [hunk 8]
```diff
@@ -22,12 +27,370 @@ Download the collection and extract it in the `data` directory, under a `collect
...
+
+### Files loading
+
+We simply open all files in the `collection` folder, loading the content in memory into a dictionnary.
+
+Since there is almost 100 000 documents, it takes aroud **15 sec** to load all the files. To avoid waiting for this operation, we save the corpus once fully loaded as a binary with `pickle` at `data/corpus.pkl`. This file takes only *161 Mo* and is loaded in **0.38 sec**.
+
+### Tokenization
+
+The data set is already tokenized, but we still this process for queries.
```
Comment (Antoine-marchais) [inline README.md:94]: typo: we still have to do this process

File: README.md [hunk 9]
```diff
@@ -22,12 +27,370 @@ Download the collection and extract it in the `data` directory, under a `collect
...
+We use the tokenization provided with `nltk`, which is more precise than a simple `split` on spaces, for numbers or `'s` for example.
+
+### Stop words
+
+We remove stop words before *lemmatization* but also after. The idea is that we want to remove all obvious useless words from the dataset before the quite *costly* lemmatization, but also remove some synonyms of stop words find with the *lemmatization*. Since the stop word removal is quite negligeable in term of computing time for the whole index creation, we think it's not a bad choice.
```
Comment (Antoine-marchais) [inline README.md:99]: typo: some synonyms of stop words **found** with the lemmatization

File: README.md [hunk 10]
```diff
@@ -22,12 +27,370 @@ Download the collection and extract it in the `data` directory, under a `collect
...
+
+### Lemmatization
+
+For a better quality of lemmatization, we use the *context* of a token in a sentence, with the `pos_tag` (*Part-Of-Speach* tag) function in `nltk`, which can indicate the grammatical use of a word : *fly* can be either a verb or a noun. The *WordNet* lemmatizer can take as argument this categorisation to give a precise result.
+
+However, since this process quite is long, so we also used the simpler version with the `SnowballStemmer` from `nltk`, which is faster.
```
Comment (Antoine-marchais) [inline README.md:105]: It must be clear that the snowball Stemmer is only a stemming process, but it is so effective that applying lemmatization + stemming with this stemmer yields almost always the same results as the stemming alone. Also specify that whichever process is chosen must also be used for the query.

File: README.md [hunk 11]
```diff
@@ -22,12 +27,370 @@ Download the collection and extract it in the `data` directory, under a `collect
...
+
+- *position index*, For each term of the collection, for each document in which the term appears, it returns both the id of the document and the positions of the term in the document.
+
+```python
+{"information": {1: [1, 19], 4: [0, 2, 5]}} # the term 'information' appears in docs 1 (at position 1 and 19) and 4 (at position at 0, 2 and 5)
+```
+
+To create an index, a cli is provided with `[src/preprocess.py](src/preprocess.py)` :
+
+```bash
```
Comment (Antoine-marchais) [inline README.md:133]: move that in usage ?

File: README.md [hunk 12]
```diff
@@ -22,12 +27,370 @@ Download the collection and extract it in the `data` directory, under a `collect
...
+
+### First analysis
+
+**Note: We're doing the time benchmark on our working machines. Therefore the resuts can vary quite a lot depending of the work actually done on the computer, process launched by the OS, etc... These benchmarks are only here to give an general idea of the steps which can take some time.**
+
+Let's benchmark a first run for our index creation, with an index of type 1, a complete lemmatization.
+
+| *Steps*    | Corpus loading | Tokenization | Stop words | Lemmatization | Stop words | Index  | Total   |
+|----------|----------------|--------------|------------|---------------|------------|--------|---------|
+| *Time (s)* | 12.38          | 80           | 67         | 679           | 49         | 2069   | **2956.38** |
```
Comment (Antoine-marchais) [inline README.md:159]: maybe specifiy the time unit for each example, this way you can show the time in minutes, which is way more understandable.

File: README.md [hunk 13]
```diff
@@ -22,12 +27,370 @@ Download the collection and extract it in the `data` directory, under a `collect
...
+| *%*        | 0.42%          | 2.71%        | 2.27%      | 22.97%        | 1.66%      | 69.98% | **100.00%** |
```
Comment (Antoine-marchais) [inline README.md:160]: Are you sure you mesured the index construction on its own ? From what I saw the decorator was on the whole index function, which includes lemmatization. If it is not on its own, then the 23% of the lemmatiation are a part of the 70% of the index ...

File: README.md [hunk 14]
```diff
@@ -22,12 +27,370 @@ Download the collection and extract it in the `data` directory, under a `collect
...
+
+Here we have two main *bottlenecks**. In the preprocessing part, we have the *lemmatization* which takes several minutes, and then the index construction is very long. Let's try to improve this two parts : processing and indexing.
+
+## Processing analysis
+
+### Introduction
```
Comment (Antoine-marchais) [inline README.md]: Add a link to another markdown here ?

File: README.md [hunk 15]
```diff
@@ -22,12 +27,370 @@ Download the collection and extract it in the `data` directory, under a `collect
...
+
+By counting the number of unique terms with low occurences in the corpus, we can see that the vast majority of the words in the corpus only occur a handfull of times, looking up some words that only occured once we found out that there were a lot of typos or references in the corpus, leading the majority of the words being very rare. As these words are very unlikely to appear in a user query, we could diminish the size of the index by removing the words with the lowest occurences.
+
+![svg](./Notebooks/processing_analysis_files/cumulative_vocab.svg)
+
+Using the previous results, we can see that 84% of the vocabulary occur less than 10 times. which shows that the stemming actually reduced quite efficiently the vocabulary among words occuring more regularly.
+
+## Index Construction
+
+Now that's the pre-processing is optimized, how can we improve the index construction.
```
Comment (Antoine-marchais) [inline README.md]: question mark ?

File: README.md [hunk 16]
```diff
@@ -22,12 +27,370 @@ Download the collection and extract it in the `data` directory, under a `collect
...
+
+Here we're making a dot product between a vector representing the query and the vector representing the dataset.
+
+**TODO**: add a few more lines
+
+### Practical querying
+
+To use and execute queries, you can use the cli provided by `[src/interface.py](src/interface.py)`.
+
+```bash
```
Comment (Antoine-marchais) [inline README.md]: move this into usage ?

File: README.md [hunk 17]
```diff
@@ -22,12 +27,370 @@ Download the collection and extract it in the `data` directory, under a `collect
...
+  --number NUMBER, -n NUMBER
+                        number of results to display (default=10)
+```
+
+The path to the index or the stats used for vectorial query are configured through *Environnement variable*, which can also be modified in `[src/config.py](src/config.py)`.
+
+For example, if you want to use a custom index, for a vectorial query with a `tf_idf_log_normalize` for both query and document weights, you could run the following query :
+
+```bash
+PATH_INDEX=/my/custom/index/path WEIGHT_QUERY=tf_idf_log_normalize WEIGHT_DOCUMENT=tf_idf_log_normalize python src/interface.py --model vectorial "cats are cute"
```
Comment (Antoine-marchais) [inline README.md]: It's a bit ugly to mix cli parameters and env variables. You can add optional cli parameters.

**Review by Antoine-marchais — COMMENTED** [review]
Most important remarks are on the report. I also added a few opinionated remarks on other parts of the code. I will be adding the vectorial query part

File: src/preprocess.py (hunk 5 above)
Comment (dixneuf19) [reply src/preprocess.py:164]: That was the idea, and then I thought : "Err, let's just write two prints"

File: Notebooks/construct_index.py (hunk 1 above)
Comment (dixneuf19) [reply Notebooks/construct_index.py]: To delete

File: src/interface.py (hunk 2 above)
Comment (dixneuf19) [reply src/interface.py:40]: removed

File: src/utils.py (hunk 4 above)
Comment (dixneuf19) [reply src/utils.py:5]: Yeah but that was useful for the benchmarks. I'll comment everything

File: src/preprocess.py (hunk 5 above)
Comment (dixneuf19) [reply src/preprocess.py:164]: commented

File: README.md (hunk 13 above)
Comment (dixneuf19) [reply README.md:160]: Yep i did this msitake at first, but then i used the tqdm part. This is an example with the very bad index type1 with a list.

**Pull request merged**